### ISSUE TITLE
SRVE8115W and SRVE8094W warnings at every request to an Krazo application

### ISSUE THREAD

**Issue by mthmulders** [OpenLiberty/open-liberty#10988]
**Describe the bug**
A sample application built with [Krazo](https://github.com/eclipse-ee4j/krazo) logs SRVE8115W and SRVE8094W warnings at every request. Krazo is built on top of JAX-RS.

Stacktrace:

    [INFO] [WARNING ] SRVE8115W: WARNING: Cannot set status. Response already committed.
    [INFO] [WARNING ] SRVE8094W: WARNING: Cannot set header. Response already committed.  Stack trace of errant attempt to set header: 
    [INFO]         at com.ibm.ws.webcontainer.srt.SRTServletResponse.setHeader(SRTServletResponse.java:1759)
    [INFO]         at [internal classes].(Unknown Source)
    [INFO]         at org.eclipse.krazo.core.ViewableWriter$DelegatingServletOutputStream.write(ViewableWriter.java:248)
    [INFO]         at java.base/java.io.OutputStream.write(OutputStream.java:157)
    [INFO]         at java.base/sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:233)
    [INFO]         at java.base/sun.nio.cs.StreamEncoder.implFlushBuffer(StreamEncoder.java:312)
    [INFO]         at java.base/sun.nio.cs.StreamEncoder.implFlush(StreamEncoder.java:316)
    [INFO]         at java.base/sun.nio.cs.StreamEncoder.flush(StreamEncoder.java:153)
    [INFO]         at java.base/java.io.OutputStreamWriter.flush(OutputStreamWriter.java:254)
    [INFO]         at java.base/java.io.PrintWriter.flush(PrintWriter.java:396)
    [INFO]         at org.eclipse.krazo.core.ViewableWriter.writeTo(ViewableWriter.java:169)
    [INFO]         at org.eclipse.krazo.core.ViewableWriter.writeTo(ViewableWriter.java:70)
    [INFO]         at org.apache.cxf.jaxrs.utils.JAXRSUtils$3.run(JAXRSUtils.java:1479)
    [INFO]         at [internal classes].(Unknown Source)

**Steps to Reproduce**
When I deploy a [sample Krazo application](https://github.com/mthmulders/openliberty-mvc) to OpenLiberty and I access it through my web browser, the OpenLiberty logs shows these messages. The sample app uses the Liberty Maven plugin so `mvn liberty:dev` should be enough to get started.

**Expected behavior**
I'd expect the warning not to be logged ;-).
The application doesn't set any headers or status itself, and as far as I can tell neither does Krazo before redirecting to the actual view.

**Diagnostic information:**
 - OpenLiberty Version: 19.0.0.12
 - Java Version: 1.8.0_212
 - server.xml configuration: pasted below, present in the sample app repo.
 
```xml
<?xml version="1.0" encoding="UTF-8"?>
<server description="defaultServer">
    <featureManager>
        <feature>beanValidation-2.0</feature>
        <feature>cdi-2.0</feature>
        <feature>jaxrs-2.1</feature>
        <feature>jsp-2.3</feature>
    </featureManager>

    <variable name="default.http.port" defaultValue="9080" />
    <variable name="default.https.port" defaultValue="9443" />

    <httpEndpoint host="*"
                  httpPort="${default.http.port}"
                  httpsPort="${default.https.port}"
                  id="defaultHttpEndpoint" />

    <webContainer disableXPoweredBy="true" />

    <webApplication location="openliberty-mvc.war" contextRoot="/" />
</server>
```

**Additional context**
Related issue in the Krazo issue tracker: [eclipse-ee4j/krazo #132](https://github.com/eclipse-ee4j/krazo/issues/132)

**Comment by andymc12** [comment]
Hi @mthmulders - I chatted with the web container team, and they suggested this trace string:
```
*=info:com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:HTTPChannel=all:GenericBNF=all
```

You can set this in the server's bootstrap.properties file (you'll need to create this file - in the same directory as the server.xml file - if it does not already exist) and then set these properties:
```
com.ibm.ws.logging.trace.specification=*=info:com.ibm.ws.webcontainer*=all:com.ibm.wsspi.webcontainer*=all:HTTPChannel=all:GenericBNF=all
com.ibm.ws.logging.max.file.size=0
```

This will enable the tracing for web container methods and will limit the trace to one (very large) file.  Once that is added to the bootstrap.properties file, restart the server and reproduce the error - at that point, we'll be looking for `SRTServletResponse.commit()`.

Thanks! Andy

**Comment by mthmulders** [comment]
Hi @andymc12, thanks for the quick reply. I've created [a gist](https://gist.github.com/mthmulders/3e8478fe84a92e46f601e4fc41d0268b#file-trace-log) with the content of trace.log. Please let me know if this is what the team expects.

**Comment by wtlucy** [comment]
Thanks for providing all this info @mthmulders - I took a look at the trace you provided.  In that trace, `flushBuffer()` is called on the response object because the Servlet's `service()` has completed - see [ServletWrapper.java#L764](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.webcontainer/src/com/ibm/ws/webcontainer/servlet/ServletWrapper.java#L764).  Without any other context, this behavior makes sense.

Interestingly, there is a webcontainer property that controls that flush: if you set the following custom property `<webContainer com.ibm.ws.webcontainer.invokeflushafterservice="false"/>` then the `SRVE8094W` will no longer occur in the example application.  A custom property is not the correct solution for this, though - we'll continue investigating.

**Comment by chkal** [comment]
@wtlucy Thanks for your reply. Let me try to explain what Eclipse Krazo is doing in detail. Maybe that helps to understand the issue.

Basically the client request is processed by a JAX-RS resource. The MessageBodyWriter than forwards the request to a JSP for creating the response. So this situation is a bit "special". However, it works fine in all other containers (Wildfly, Glassfish, Tomcat, etc).

**Comment by andymc12** [comment]
@mthmulders Using your app I reproduced the problem.  As I understand it, everything works functionally, but there is the nasty warning messages in the logs that we would like to avoid.  Is that right?

I also tried @wtlucy 's suggestion of using `<webContainer com.ibm.ws.webcontainer.invokeflushafterservice="false"/>` - with this setting, everything worked, and I had no nasty warning messages.  

If the MVC app is the only thing running in the Liberty server instance, then this is probably the best solution.  Essentially this setting prevents the web container from flushing the stream after the servlet's service method has exited.  This should be fine in cases where the application code flushes the stream or in cases where the request is forwarded (as in this case).

I have been chatting with @wtlucy about a more targeted approach that would allow Krazo to set a per-request setting that does the same thing - that way, Krazo requests (which should all end with a forwarded request) would avoid the extra flush call, but all other requests would not.

First, can you confirm that this setting works for you?

Second, does the proposed solution sound good to you (and to @chkal)?  If so, we should use this issue to drive in that solution.

Thanks, Andy

**Comment by chkal** [comment]
@andymc12 Thanks a lot for helping us with this :-)

I guess such a solution would definitely work. However, I still don't fully understand why this warning is logged and IMO what we are doing is fully compliant with the spec. Therefore, I would really like to understand what exactly is causing this behavior.

Let me try to explain in more detail what is happening for such an MVC web request.

1. The browser request hits the container and is processed by JAX-RS + MVC
2. It may be possible that as few headers are set on the reponse at this point.
3. The Krazo JAX-RS MBW invokes `RequestDispatcher.forward()` to forward the request to a JSP
4. The corresponding JSP is rendered (and will most likely commit the response and/or flush buffers)
5. After the forward has completed, Krazo flushes the response writer again. This seems to trigger the warning in the stacktrace.

So IMO it is fine for step 4 to flush and commit the response. And IMO it is also fine for Krazo to flush the stream again in step 5. What I don't understand is that flushing the stream in step 5 also leads to a situation where Liberty calls `response.setHeader()` (see stacktrace in the issue description) which obviously fails as the response is already committed.

To sum up: Why is Liberty trying to set a response header when the application flushes the response? From the detailed log it looks like Liberty tries to set a `Date` header at this point. Is that expected behavior?

**Comment by wtlucy** [comment]
@chkal thanks for the detailed explanation - it helped to understand the scenario.  The basic problem, as you said, was that the call to `flush()` (step 5) caused the Liberty CXF implementation attempt to write out headers.  Since the app code (Krazo) wasn't directly responsible for the late header write attempt, we decided that the warnings were not correct.  

In #11070 Andy updated the CXF code to check the response status before writing any headers, so the `SRVE8115W` and `SRVE8094W` warnings will no longer occur on this path.  `mthmulders` can you verify this issue is resolved?  https://openliberty.io/downloads/#development_builds

**Comment by mthmulders** [comment]
I've downloaded the OpenLiberty build dated 2020-03-03, 08:10 and can confirm that the SRVE8115W and SRVE8094W warnings no longer occur!

Excellent work, thanks @andymc12 and @chkal for this close cooperation!

**Comment by chkal** [comment]
@mthmulders Thanks for the confirmation!
@andymc12 @wtlucy Thanks a lot for fixing this issue!

**Comment by andymc12** [comment]
No problem!  Thanks for reporting this - I'm looking forward to seeing the blog post on how to use Krazo (MVC) on Open Liberty!

**Comment by mthmulders** [comment]
It has been there for 1,5 year...: https://maarten.mulders.it/2018/09/a-first-look-at-mvc-1.0/
